ui/app/hooks/useRetryTransaction.js

```diff
@@ -0,0 +1,110 @@
+import { useDispatch } from 'react-redux'
+import { useMemo, useCallback, useEffect, useState } from 'react'
+import * as actions from '../store/actions'
+import {
+  fetchBasicGasAndTimeEstimates as fetchBasicGasAndTimeEstimatesAction,
+  fetchGasEstimates as fetchGasEstimatesAction,
+  setCustomGasPriceForRetry as setCustomGasPriceForRetryAction,
+  setCustomGasLimit as setCustomGasLimitAction,
+} from '../ducks/gas/gas.duck'
+import { TOKEN_METHOD_TRANSFER } from '../helpers/constants/transactions'
+import { increaseLastGasPrice } from '../helpers/utils/confirm-tx.util'
+import { useMetricEvent } from './useMetricEvent'
+import { useMethodData } from './useMethodData'
+
+
+/**
+ * useRetryTransaction
+ *
+ * Provides a reusable hook that, given a transactionGroup, will return
+ * whether or not the transaction meets the criteria to be retried with
+ * higher gas (sped up), and a method for beginning the retry process
+ * @param {Object} transactionGroup - the transaction group
+ * @param {boolean} isEarliestNonce - indicates if this transaction is the earliest in history
+ * @return {[boolean, Function]}
+ */
+export function useRetryTransaction (transactionGroup, isEarliestNonce = false) {
+  const { primaryTransaction, transactions, initialTransaction, hasRetried } = transactionGroup
+  const [earliestTransaction = {}] = transactions
+  const { submittedTime } = earliestTransaction
+  const gasPrice = primaryTransaction.txParams?.gasPrice
+  const methodData = useMethodData(primaryTransaction.txParams?.data)
+  const trackMetricsEvent = useMetricEvent(({
+    eventOpts: {
+      category: 'Navigation',
+      action: 'Activity Log',
+      name: 'Clicked "Speed Up"',
+    },
+  }))
+  const dispatch = useDispatch()
+  const { fetchBasicGasAndTimeEstimates, fetchGasEstimates, setSelectedToken, retryTransaction } = useMemo(() => ({
+    fetchBasicGasAndTimeEstimates: () => dispatch(fetchBasicGasAndTimeEstimatesAction()),
```

Comment:
It prevents the other three functions from being created every render, only recreating them when the retryTransaction method must also be constructed. This is close to the behavior of mapDispatchToProps. ``

Reply:
Technically it does re-create the functions every render though - it just doesn't use them. The only performance advantage here is in preserving the referential equality of this function in child components (i.e. preventing re-renders), but I don't think we _use_ these callbacks in that matter, so... this is more work for no gain, I think? Unless I'm misunderstanding something here.

Here's the explanation I was thinking of: https://kentcdodds.com/blog/usememo-and-usecallback

Reply:
Separately from that, this caught my eye originally because of how these functions were grouped together. It would seem better to have them in separate `useCallback` hooks, so that each is only re-created when necessary instead of them _all_ being recreated when any dep changes.

Reply:
@Gudahtt 🤦‍♂️ just when you think you know something. Thanks for sharing the article. The truth is wrapping those three other functions is just mimicking the mapDispatchToProps... there's almost nothing to gain from actually creating a new function that just calls dispatch because I'm not passing the functions anywhere.. i could just call dispatch directly where I am calling these methods in the code... which means even less memory allocation because the selectors are instantiated once, dispatch is allocated by useDispatch...

Reply:
Lemme know what you think of the new flow @Gudahtt -- just using dispatch directly instead of creating the intermediary functions

Reply:
Works for me!